onnxscript/function_libs/torch_aten/ops/core.py

```diff
@@ -1075,10 +1075,14 @@ def aten_constant_pad_nd(self: TensorType, pad: INT64, value: float = 0.0) -> Te
     raise NotImplementedError()
 
 
-def aten_contiguous(self: TensorType, memory_format: str = "contiguous_format") -> TensorType:
+@torch_op("aten::contiguous", trace_only=True)
+def aten_contiguous(self: TTensor, memory_format: str = "contiguous_format") -> TTensor:
     # contiguous(Tensor(a) self, *, MemoryFormat memory_format=contiguous_format) -> Tensor(a)
 
-    raise NotImplementedError()
+    if memory_format in ["contiguous_format", "preserve_format"]:
```

Comment:
I wonder if we can always return self? This way we reduce the amount of trace_only ops

Reply:
We can leave such data validation work out of aten lib functions. If the string is wrong, when use inference the ONNX graph, ort is supposed to throw an error.

Reply:
good point as @fatcat-z says. thanks for pointing out.

Reply:
Offline discussed with Jay, finally we think we need break the pipeline at this time, because ort doesn't support the other 2 format string.  
If we always return op.Identity(self), but when the memory_format in ["channels_last", "channels_last_3d"] in torch model, this will leading value errors in onnx model.

Reply:
@justinchuby 
Yes, this is different than what I thought before. In current case, the value of memory_format won't be passed to ORT meaning ORT would not know the memory_format might be an unsupported one (channels_last/channels_last_3d). And this will cause a wrong inference result and it could only be found when we check this value in ONNX graph which is unfriendly.

Based on this, probably the best solution is to reminder users during the export by raising an exception or printing a warning.

Thoughts?

Reply:
I think that’s ok. The functions will be much clearer if we restrict them to not raise any errors. Also for consistency, because only the trace only functions can raise errors. 

I would leave a todo in-line for this case, and handle them in the future when we can annotate constraints for argument, as part of the function meta data structure

Reply:
@gramalingam 

Do you have any suggestions for such case?

Reply:
I think it is good to make sure we don't export models that will have a different behavior. It is better that users see failure at export-time instead of getting an erroneous model exported. I don't quite understand Justin's last comment. Are you suggesting some other mechanism for catching such "unsupported cases"? It seems to me an exception, trace-only mode, etc. seems preferable to a todo comment.

Reply:
I don't understand the op spec in aten. I guess it is supposed to alter the layout representation in some cases, but not others? May be I am reading too much into the names, but I am puzzled how both "contiguous_format" and "preserve_format" get mapped to identity.

Reply:
> It is better that users see failure at export-time instead of getting an erroneous model exported

I agree that we should get failures at export-time. I think it would be nice to keep the functions purely descriptive (logic, signature, expected input) so that trace_only and onnxfunctions can be seem the same. I am playing with the idea that we annotate the arguments (via Annotated, or via a decorator) to specify any additional constraints which the exporter can then pick up and handle errors accordingly. The TODO comment was suggested for adding this annotation later when we implement this mechanism.

> I am puzzled how both "contiguous_format" and "preserve_format" get mapped to identity.

That's a good point. Maybe the wise thing to do for this op in particular is to drop it in the FX exporter and never send it to ONNX? Does onnx have a notion of memory format?

Reply:
So probably the best way for now is:
1) Keep raising this exception if the value of memory_format is invalid for ONNX.
2) Add a TODO comment to mention Justin's idea so we might come back once that idea has been proved is feasible.

Reply:
> > It is better that users see failure at export-time instead of getting an erroneous model exported
> 
> I agree that we should get failures at export-time. I think it would be nice to keep the functions purely descriptive (logic, signature, expected input) so that trace_only and onnxfunctions can be seem the same. I am playing with the idea that we annotate the arguments (via Annotated, or via a decorator) to specify any additional constraints which the exporter can then pick up and handle errors accordingly. The TODO comment was suggested for adding this annotation later when we implement this mechanism.
> 
> > I am puzzled how both "contiguous_format" and "preserve_format" get mapped to identity.
> 
> That's a good point. Maybe the wise thing to do for this op in particular is to drop it in the FX exporter and never send it to ONNX? Does onnx have a notion of memory format?

Current PyTorch exporter could export it so it'd better we don't lose it.

Reply:
> So probably the best way for now is:
> 
> 1. Keep raising this exception if the value of memory_format is invalid for ONNX.
> 2. Add a TODO comment to mention Justin's idea so we might come back once that idea has been proved is feasible.

ok

Reply:
add below comment:
'''
# TODO: find some fundamental solutoin to solve this kind of error
'''